java/refactoring.java/src/org/netbeans/modules/refactoring/java/plugins/InnerToOuterTransformer.java

```diff
@@ -582,44 +586,80 @@ private ClassTree refactorInnerClass(ClassTree innerClass) {
                         Set<Modifier> modifiers = EnumSet.noneOf(Modifier.class);
                         modifiers.addAll(newConstructor.getModifiers().getFlags());
                         modifiers.remove(Modifier.PRIVATE);
                         newConstructor = make.Constructor(
                                 make.Modifiers(modifiers,newConstructor.getModifiers().getAnnotations()),
-                                newConstructor.getTypeParameters(), 
+                                newConstructor.getTypeParameters(),
                                 newConstructor.getParameters(),
                                 newConstructor.getThrows(),
                                 block);
 
-                        newInnerClass = make.removeClassMember(newInnerClass, m);
-                        genUtils.copyComments(m, newConstructor, true);
-                        genUtils.copyComments(m, newConstructor, false);
+                        newInnerClass = make.removeClassMember(newInnerClass, ctor);
+                        genUtils.copyComments(ctor, newConstructor, true);
+                        genUtils.copyComments(ctor, newConstructor, false);
                         newInnerClass = genUtils.insertClassMember(newInnerClass, newConstructor);
                     }
                 }
             }
         }
-        
+
         if(innerClass != newInnerClass) {
             genUtils.copyComments(innerClass, newInnerClass, true);
             genUtils.copyComments(innerClass, newInnerClass, false);
         }
         return newInnerClass;
     }
 
+    private ClassTree refactorInnerRecord(ClassTree innerRecord) {
+        refactoring.setInnerIsRecord(true);
+        ClassTree newInnerRecord = innerRecord;
+        List<? extends Tree> members = newInnerRecord.getMembers();
+        int insertAt=0;
+        for (Tree member : members) {
+            Tree.Kind kind = member.getKind();
+//            System.out.println("member = " + kind + "" + member);
+            
+            switch (kind) {
+               // The fields, except the static, should loose private final modifiers
+               // so that they appear without these modifiers in the record header
+                case VARIABLE:
+                    VariableTree vt = (VariableTree) member;
+                    ModifiersTree mt = vt.getModifiers();
+                    if (mt.getFlags().contains(Modifier.STATIC)) {
+                        continue;
+                    }
+                    var mtype = vt.getType();
+                    var mName = vt.getName();
+                    var mods = make.Modifiers(EnumSet.noneOf(Modifier.class));
+                    var newMember = make.RecordComponent(mods, mName, mtype);
+                    newInnerRecord = make.removeClassMember(newInnerRecord, member);
+                    newInnerRecord = make.insertClassMember(newInnerRecord, insertAt, newMember);
```

Comment:
i am not a fan of `var` since it hides the type only so that the IDE can render it next to it again as inline hint (-> nothing is gained and it makes review more difficult). But I think i can figure out what the types are here so I let it slide ;)